source/common/router/filter_config_discovery_impl.cc

```diff
@@ -0,0 +1,191 @@
+        "Unexpected number of resources in FilterConfigDS response: {}", resources.size()));
+  }
+  const auto& filter_config = dynamic_cast<const envoy::config::core::v3::TypedExtensionConfig&>(
+      resources[0].get().resource());
+  if (filter_config.name() != filter_config_name_) {
+    throw EnvoyException(fmt::format("Unexpected resource name in FilterConfigDS response: {}",
+                                     filter_config.name()));
+  }
+  auto& factory = Envoy::Config::Utility::getAndCheckFactory<
+      Server::Configuration::NamedHttpFilterConfigFactory>(filter_config);
+  // Ensure that the factory is valid in the filter chain context.
+  for (auto* provider : filter_config_providers_) {
+    provider->validateConfig(factory);
+  }
+  ProtobufTypes::MessagePtr message = Envoy::Config::Utility::translateAnyToFactoryConfig(
+      filter_config.typed_config(), validator_, factory);
+  Http::FilterFactoryCb factory_callback =
+      factory.createFilterFactoryFromProto(*message, stat_prefix_, factory_context_);
+  ENVOY_LOG(debug, "Updating filter config {}", filter_config_name_);
+  for (auto* provider : filter_config_providers_) {
+    provider->onConfigUpdate(factory_callback, version_info);
+  }
+}
+
+void FilterConfigSubscription::onConfigUpdate(
+    const std::vector<Envoy::Config::DecodedResourceRef>& added_resources,
+    const Protobuf::RepeatedPtrField<std::string>& removed_resources, const std::string&) {
+  if (!removed_resources.empty()) {
+    ENVOY_LOG(error,
+              "Server sent a delta FilterConfigDS update attempting to remove a resource (name: "
+              "{}). Ignoring.",
+              removed_resources[0]);
+  }
+  if (!added_resources.empty()) {
+    onConfigUpdate(added_resources, added_resources[0].get().version());
+  }
+}
+
+void FilterConfigSubscription::onConfigUpdateFailed(Envoy::Config::ConfigUpdateFailureReason reason,
+                                                    const EnvoyException*) {
+  ENVOY_LOG(debug, "Updating filter config {} failed due to {}", filter_config_name_, reason);
+  // Make sure to make progress in case the control plane is temporarily failing.
+  init_target_.ready();
+}
+
+FilterConfigSubscription::~FilterConfigSubscription() {
+  // If we get destroyed during initialization, make sure we signal that we "initialized".
+  init_target_.ready();
+  // Remove the subscription from the provider manager.
+  filter_config_provider_manager_.subscriptions_.erase(subscription_id_);
+}
+
+std::shared_ptr<FilterConfigSubscription> FilterConfigProviderManagerImpl::getSubscription(
+    const envoy::config::core::v3::ConfigSource& config_source, const std::string& name,
+    Server::Configuration::FactoryContext& factory_context, const std::string& stat_prefix) {
+  // FilterConfigSubscriptions are unique based on their config source and filter config name
+  // combination.
+  const std::string subscription_id = absl::StrCat(MessageUtil::hash(config_source), ".", name);
```

Comment:
This is safe, even in the presence of collisions, since ultimately the returned subscription has the actual name embedded in it? I know we do this elsewhere, just trying to convince myself that this is correct.

Reply:
Hash collisions are not safe since two config sources may not have identical resources for a name. But hash of config source proto is commonly used.

Reply:
My main thinking here, which is more general than this PR, is that in a multi-tenant world where one tenant has the ability to create its own names, it might be fairly easy to engineer a hash collision, since we don't have crypto-strength hashing here. If you reckon this sounds right, I can just file an issue.

Reply:
Yeah, this is a possible attack with a rogue control plane (e.g. send an LDS update with a made up RDS config source that hijacks another LDS/RDS reference). I've seen other instances of hashing xDS payloads, so it might extend to cases where a control plane can subsume another resource by engineering a hash collision.

Reply:
Created https://github.com/envoyproxy/envoy/issues/11967 to track.

Reply:
Can you add this as a TODO here?

Reply:
Done.